In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import warnings
import sklearn
from sklearn import tree
from sklearn.naive_bayes import GaussianNB

In [2]:
trainx=pd.read_csv("./HeartDisease.csv")

In [3]:
print(trainx.head())

   ID  Age  Sex  cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0   1   28    1   2     130.0  132.0  0.0      2.0    185.0    0.0      0.0   
1   2   29    1   2     120.0  243.0  0.0      0.0    160.0    0.0      0.0   
2   3   29    1   2     140.0    NaN  0.0      0.0    170.0    0.0      0.0   
3   4   30    0   1     170.0  237.0  0.0      1.0    170.0    0.0      0.0   
4   5   31    0   2     100.0  219.0  0.0      1.0    150.0    0.0      0.0   

   num    Place  
0    0  Hungary  
1    0  Hungary  
2    0  Hungary  
3    0  Hungary  
4    0  Hungary  


In [4]:
trainx.describe()

ID         Age         Sex          cp    trestbps        chol  \
count  457.000000  457.000000  457.000000  457.000000  424.000000  431.000000   
mean   229.000000   50.962801    0.796499    3.155361  131.488208  200.048724   
std    132.068795    9.219588    0.403043    0.943548   18.829454  113.463794   
min      1.000000   28.000000    0.000000    1.000000   80.000000    0.000000   
25%    115.000000   45.000000    1.000000    2.000000  120.000000  171.000000   
50%    229.000000   52.000000    1.000000    3.000000  130.000000  220.000000   
75%    343.000000   57.000000    1.000000    4.000000  140.000000  269.500000   
max    457.000000   75.000000    1.000000    4.000000  200.000000  603.000000   

              fbs     restecg     thalach       exang     oldpeak         num  
count  403.000000  455.000000  424.000000  424.000000  421.000000  457.000000  
mean     0.138958    0.356044  134.910377    0.358491    0.622090    0.459519  
std      0.346332    0.590529   24.251177    0.480124    0.922115    0.498905  
min      0.000000    0.000000   69.000000    0.000000   -2.600000    0.000000  
25%      0.000000    0.000000  118.000000    0.000000    0.000000    0.000000  
50%      0.000000    0.000000  135.000000    0.000000    0.000000    0.000000  
75%      0.000000    1.000000  150.000000    1.000000    1.000000    1.000000  
max      1.000000    2.000000  190.000000    1.000000    5.000000    1.000000

Pasamos la variable de lugar a nuerica. 0 para hungary, 1 para VA y 2 para otros

In [24]:
train = trainx.copy()
train["Place"][train["Place"] == "Hungary"] = 0
train["Place"][train["Place"] == "VA"] = 1
train["Place"][(train["Place"] != "VA") | (train["Place"] != "Hungary")] = 2


/home/dacer/anaconda3/envs/proyecto/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/home/dacer/anaconda3/envs/proyecto/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/home/dacer/anaconda3/envs/proyecto/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/ind

In [25]:
train.describe()


ID         Age         Sex          cp    trestbps        chol  \
count  457.000000  457.000000  457.000000  457.000000  424.000000  431.000000   
mean   229.000000   50.962801    0.796499    3.155361  131.488208  200.048724   
std    132.068795    9.219588    0.403043    0.943548   18.829454  113.463794   
min      1.000000   28.000000    0.000000    1.000000   80.000000    0.000000   
25%    115.000000   45.000000    1.000000    2.000000  120.000000  171.000000   
50%    229.000000   52.000000    1.000000    3.000000  130.000000  220.000000   
75%    343.000000   57.000000    1.000000    4.000000  140.000000  269.500000   
max    457.000000   75.000000    1.000000    4.000000  200.000000  603.000000   

              fbs     restecg     thalach       exang     oldpeak         num  
count  403.000000  455.000000  424.000000  424.000000  421.000000  457.000000  
mean     0.138958    0.356044  134.910377    0.358491    0.622090    0.459519  
std      0.346332    0.590529   24.251177    0.480124    0.922115    0.498905  
min      0.000000    0.000000   69.000000    0.000000   -2.600000    0.000000  
25%      0.000000    0.000000  118.000000    0.000000    0.000000    0.000000  
50%      0.000000    0.000000  135.000000    0.000000    0.000000    0.000000  
75%      0.000000    1.000000  150.000000    1.000000    1.000000    1.000000  
max      1.000000    2.000000  190.000000    1.000000    5.000000    1.000000

In [26]:
print(train.head())

   ID  Age  Sex  cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0   1   28    1   2     130.0  132.0  0.0      2.0    185.0    0.0      0.0   
1   2   29    1   2     120.0  243.0  0.0      0.0    160.0    0.0      0.0   
2   3   29    1   2     140.0    NaN  0.0      0.0    170.0    0.0      0.0   
3   4   30    0   1     170.0  237.0  0.0      1.0    170.0    0.0      0.0   
4   5   31    0   2     100.0  219.0  0.0      1.0    150.0    0.0      0.0   

   num Place  
0    0     2  
1    0     2  
2    0     2  
3    0     2  
4    0     2  


### Analizamos como rellenar los datos faltantes
- en el colesterol (chol) es interesante pensar en el promedio, 
    aunque se me ocurren muchas maneras, como poner el promedio de similares.
- en la presion arterial con la que ingresaron (trestbps) es interesante utilizar promedio
    aunque plantear el uso de similitudes se debe considerar.
- en el campo que indica si si tiene el azucar en la sangre > 120 mg/dl (fbs) se debe considerar que 1 es para true
    y 0 es para false, sin embargo no es un campo que podamos rellenar a la ligera
    por lo tanto se debe considerar si se elimina.
- Los resultados electrocardiográficos en reposo (restecg) son Value 0: normal, Value 1: con anormalidad de la onda ST-T (inversiones de la onda T y / o elevación o depresión del ST> 0.05 mV) y Valor 2: muestra hipertrofia ventricular izquierda probable o definitiva según los criterios de Estes. Por todo lo anterior rellenarlos no es factible con el promedio.
- La frecuencia cardíaca máxima alcanzada (thalach) es otro atributo que se puede relenar con promedios pero
    considerando que esto es un principal sintoma de afeciones cardiacas, por lo que recomiendo
    utilizar el promedio pero de los que poseen afecciones en caso de afectado, o promedio de los que no en caso
    de ser una persona que no se le encontro afeciones cardiacas.
- La depresión del ST inducida por el ejercicio relativo al descanso (oldpeak) tiene como valores Valor 1: pendiente ascendente - Valor 2: plano - Valor 3: pendiente descendente. este valor no es factible rellenarlo
- La angina inducida por ejercicio (1 = sí; 0 = no) (exang) es otro valor que conviene analizar la forma de rellenarlo, ya que aplicar un simple promedio no es valido.
- El valor objetivo num que determina la existencia de problemas patologias cardiacas tbn se encuentra sin rellenar

    


partimos en 80 20

In [27]:
from sklearn.model_selection import train_test_split


# vamos a definir los objetivos en target y los predictores
features = train[["Age","Sex","cp", "trestbps", "thalach", "exang","chol","fbs","restecg","oldpeak"]].values
target = train["num"].values


# Split the data into train and test
trainX, testX, trainY, testY = train_test_split(features, target, test_size=0.2)
print(trainX.shape, trainY.shape)
print(testX.shape, testY.shape)

(365, 10) (365,)
(92, 10) (92,)


## Entrenamiento
Aqui se debe considerar que la predición no se pude realizar si no se rellenan todos los datos, 
por tal motivo esta parte estara pendiente hasta discutir el relleno de datos del corazon

In [28]:
gnb = GaussianNB()

model2 = gnb.fit(trainX, trainY)

#Print the score on the train data
print(model2.score(trainX, trainY))
#Print the score on the test data
print(model2.score(testX, testY))

ValueError: Input contains NaN, infinity or a value too large for dtype('float64').